raft_node.py

```diff
@@ -0,0 +1,595 @@
+                case FollowerState(
+                    leader=leader
+                ) if leader == rpc_request.requester_id:
+                    match rpc_request:
+                        case AppendEntriesRequest():
+                            # Reset the heartbeat timer
+                            run_in_background(self._wait_for_next_heartbeat())
+
+        return f(self, rpc_request)
+
+    return wrapper
+
+
+@dataclass(slots=True)
+class Node:
+    config: Config
+    _state: NodeState = field(default_factory=FollowerState)
+    _current_term: int = field(init=False)
+    _voted_for: int | None = field(init=False)
+    _log: Log = field(init=False)
+    _raft_rpc_server: raft_server.Server = field(init=False)
+    _cluster_rpc_server: cluster_server.Server = field(init=False)
+    _raft_rpc_clients: dict[int, Client] = field(init=False)
+
+    def __post_init__(self: Self) -> None:
+        self._raft_rpc_server = raft_server.Server(
+            port=self.config.raft_server_port,
+            request_vote_handler=self._request_vote_handler,
+            append_entries_handler=self._append_entries_handler,
+        )
+
+        cluster_servicer: cluster_server.Servicer = cluster_server.Servicer(
+            get_value_handler=self._get_value_handler,
+            propose_mapping_handler=self._propose_mapping_handler,
+        )
+        self._cluster_rpc_server = cluster_server.Server(
+            port=self.config.cluster_server_port, servicer=cluster_servicer
+        )
+
+        self._raft_rpc_clients = {
+            other_port: Client(port=other_port)
+            for other_port in self.config.all_raft_server_ports
+            if other_port != self.config.raft_server_port
+        }
+
+    def _print(self: Self, msg: str) -> None:
+        print(f"Node {self.config.raft_server_port}: {msg}")
+
+    async def _load_persistent_state(self: Self) -> None:
+        self._current_term, self._voted_for = -1, None
+        if os.path.isfile(self.config.persistent_state_file):
+            with open(self.config.persistent_state_file, "r") as f:
+                self._current_term, self_voted_for = json.load(f)
+                self._print(
+                    f"Initialized persistent state from disk: "
+                    f"term {self._current_term}, "
+                    f"voted_for {self._voted_for}"
+                )
+        else:
+            self._print("No persistent state file. Initializing term to -1")
+
+        self._log = await Log.create(self.config.log_file)
+        self._print(f"Initial state machine: {self._log._cur_state}")
+        self._print(f"Last index and term: {self._log.last_index_and_term()}")
+
+    async def _update_persistent_state(self: Self) -> None:
+        async with aiofiles.open(self.config.persistent_state_file, "w") as f:
+            contents: str = json.dumps((self._current_term, self._voted_for))
+            await f.write(contents)
+            await f.flush()
```

Comment:
Flushing seems unnecessary? I'd assume that aiofiles does something reasonable when closing the file.